General cleanup of TransformOperations

weinig opened the pull request:
#### 747e839981525e26209151149e110e2ab07b4ded
<pre>
General cleanup of TransformOperations
<a href="https://bugs.webkit.org/show_bug.cgi?id=274851">https://bugs.webkit.org/show_bug.cgi?id=274851</a>

Reviewed by Darin Adler.

Makes some small improvements to the TransformOperations type:

- Use Ref instead of RefPtr in operations vector.
- Made immutable. All construction already happened in one shot in the builder
  or blend functions, or was a full replacement, so this just meant removing
  non-const accessors and removing the clear() function.
- Expose iterators directly to avoid clients accessing the underlying vector.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h: Added.
    - Split SharedPrimitivesPrefix class into its own file.

* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
* Source/WebCore/platform/graphics/transforms/TransformOperations.h:
    - Convert to use a Vector&lt;Ref&lt;TransformOperation&gt;&gt; for the implementation,
      remove all mutating functionality, and add some helpers to make call sites
      more clear.
    - Swap the size and matrix arguments in apply to match individual transforms.

* Source/WebCore/animation/BlendingKeyframes.cpp:
    - Update to use new direct iteration.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
    - Use an explicit construction of a Vector&lt;Ref&lt;TransformOperation&gt;&gt; to build the
      TransformOperations, this allows for an exactly sized result.
    - Specialize the property wrapper for TransformOperations to allow moving in the
      result to the setter, avoiding a Vector copy.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::checkForMatchingTransformFunctionLists):
(WebCore::KeyframeEffect::computeTransformedExtentViaTransformList const):
(WebCore::containsRotation): Deleted.
    - Use new helpers to simplify use.

* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Made TransformSyntaxValue store a Ref, allowing for removing some
      null checks that were never possibly null.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
    - Update to use new direct iteration.

* Source/WebCore/css/TransformFunctions.cpp:
    - Use an explicit construction of a Vector&lt;Ref&lt;TransformOperation&gt;&gt; to build the
      TransformOperations using a failable generator constructor.

* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
    - Use initialization syntax in copy constructor to avoid unnecessary
      re-initialization. Use new helpers to simplify use.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
    - Use new helpers to simplify use.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::hasBig3DRotation):
    - Update for new, more clear, name of SharedPrimitivesPrefix.

* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.h:
    - Mark final overrides final, and make them public so callers that have
      the specific type can avoid a virtual call (RenderStyle makes use of
      this below).

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
    - Use an explicit construction of a Vector&lt;Ref&lt;TransformOperation&gt;&gt; to build the
      TransformOperations. By using the initialization_list constructor, the vector
      is perfectly sized.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::transformMatrixForProperty const):
    - Update for swapped apply arguments.

* Source/WebCore/rendering/mathml/RenderMathMLOperator.cpp:
    - Remove unused includes.

* Source/WebCore/rendering/style/RenderStyle.cpp:
    - Use existing accessors rather than writing full path out each time.
    - Use auto* rather than TransformOperation* to allow for de-virtualization
      of the calls to apply().

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
    - Take TransformOperations by r-value to avoid all sets causing a copy. Like
      with Length, StyleBuilder will do an explicit copy when forwarding.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::forwardInheritedValue):
    - Add overload of forwardInheritedValue for TransformOperations that copies
      before moving into RenderStyle.

* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasTransform const):
    - Call isEmpty() directly on the container.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Update to directly access operations via the member variable, necessary
      now that there is no operations() function.

* Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
(Nicosia::applyTransformAnimation):
    - Update for swapped apply arguments.

Canonical link: <a href="https://commits.webkit.org/279618@main">https://commits.webkit.org/279618@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a2b25989df57ed91b6114732e9b889fc90b0076

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53999 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33371 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6530 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57275 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4724 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56301 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4615 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43720 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3121 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56096 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46725 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24861 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28418 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4047 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2873 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50078 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4247 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58869 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29179 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4333 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51136 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30365 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50479 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11763 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31319 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30142 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -4353,22 +4372,24 @@ static std::optional<CSSCustomPropertyValue::SyntaxValueList> blendSyntaxValueLi
-            return transformOperations;
+        auto transformOperationsFromSyntaxValueList = [](const CSSCustomPropertyValue::SyntaxValueList& list) {
+            return TransformOperations {
+                list.values.map([](auto& syntaxValue) -> Ref<TransformOperation> {
```
Does the `Ref<TransformOperation>` return value need to be here explicitly?

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -4353,22 +4372,24 @@ static std::optional<CSSCustomPropertyValue::SyntaxValueList> blendSyntaxValueLi
+        auto transformOperationsFromSyntaxValueList = [](const CSSCustomPropertyValue::SyntaxValueList& list) {
+            return TransformOperations {
+                list.values.map([](auto& syntaxValue) -> Ref<TransformOperation> {
+                    ASSERT(std::holds_alternative<CSSCustomPropertyValue::TransformSyntaxValue>(syntaxValue));
```
Why do we need this assertion? Doesn't `std::get` effectively also assert this? If not, should we add a non-standard alternative that includes the assertion?

darinadler commented on Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
```diff
@@ -40,29 +40,22 @@
 namespace WebCore {
 
 AcceleratedEffectValues::AcceleratedEffectValues(const AcceleratedEffectValues& src)
```
Is there any trick we can do to get this to be generated. Seems annoying to have to write it out just for those `copyRef` calls.

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformOperations.h:
```diff
@@ -35,90 +36,91 @@ struct BlendingContext;
+
+    TransformOperations clone() const
+    {
+        return TransformOperations { m_operations.map([](const auto& op) { return op->clone(); }) };
```
Can we move these multi-line function bodies out of the class? If they can’t be in the `.cpp` file, at least we could put them below the class, so it’s easier to see the class’s interface.

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h:
```diff
@@ -0,0 +1,55 @@
+
+class TransformOperations;
+
+// This class  is used to find a shared prefix of transform function primitives (as
```
Extra space here between "class" and "is".

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h:
```diff
@@ -0,0 +1,55 @@
+class TransformOperationsSharedPrimitivesPrefix final {
+public:
+    TransformOperationsSharedPrimitivesPrefix() = default;
+    ~TransformOperationsSharedPrimitivesPrefix() = default;
```
I think you can just omit these, and we should get the same behavior.

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.cpp:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 1999 Antti Koivisto (koivisto@kde.org)
+ * Copyright (C) 2004, 2005, 2006, 2007, 2008 Apple Inc. All rights reserved.
```
Probably too much copyright for this one function; would be nice to get it more precise.

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.cpp:
```diff
@@ -0,0 +1,56 @@
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
```
Should be the other license, I think.

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h:
```diff
@@ -0,0 +1,55 @@
+ *           (C) 2000 Antti Koivisto (koivisto@kde.org)
+ *           (C) 2000 Dirk Mueller (mueller@kde.org)
+ * Copyright (C) 2003, 2005, 2006, 2007, 2008 Apple Inc. All rights reserved.
+ * Copyright (C) 2006 Graham Dennis (graham.dennis@gmail.com)
```
Too much copyright for this one class.

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h:
```diff
@@ -0,0 +1,55 @@
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
```
Should be the other license, I think.

Commit:
General cleanup of TransformOperations
https://bugs.webkit.org/show_bug.cgi?id=274851

Reviewed by Darin Adler.

Makes some small improvements to the TransformOperations type:

- Use Ref instead of RefPtr in operations vector.
- Made immutable. All construction already happened in one shot in the builder
  or blend functions, or was a full replacement, so this just meant removing
  non-const accessors and removing the clear() function.
- Expose iterators directly to avoid clients accessing the underlying vector.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h: Added.
    - Split SharedPrimitivesPrefix class into its own file.

* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
* Source/WebCore/platform/graphics/transforms/TransformOperations.h:
    - Convert to use a Vector<Ref<TransformOperation>> for the implementation,
      remove all mutating functionality, and add some helpers to make call sites
      more clear.
    - Swap the size and matrix arguments in apply to match individual transforms.

* Source/WebCore/animation/BlendingKeyframes.cpp:
    - Update to use new direct iteration.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
    - Use an explicit construction of a Vector<Ref<TransformOperation>> to build the
      TransformOperations, this allows for an exactly sized result.
    - Specialize the property wrapper for TransformOperations to allow moving in the
      result to the setter, avoiding a Vector copy.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::checkForMatchingTransformFunctionLists):
(WebCore::KeyframeEffect::computeTransformedExtentViaTransformList const):
(WebCore::containsRotation): Deleted.
    - Use new helpers to simplify use.

* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Made TransformSyntaxValue store a Ref, allowing for removing some
      null checks that were never possibly null.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
    - Update to use new direct iteration.

* Source/WebCore/css/TransformFunctions.cpp:
    - Use an explicit construction of a Vector<Ref<TransformOperation>> to build the
      TransformOperations using a failable generator constructor.

* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
    - Use initialization syntax in copy constructor to avoid unnecessary
      re-initialization. Use new helpers to simplify use.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
    - Use new helpers to simplify use.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::hasBig3DRotation):
    - Update for new, more clear, name of SharedPrimitivesPrefix.

* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.h:
    - Mark final overrides final, and make them public so callers that have
      the specific type can avoid a virtual call (RenderStyle makes use of
      this below).

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
    - Use an explicit construction of a Vector<Ref<TransformOperation>> to build the
      TransformOperations. By using the initialization_list constructor, the vector
      is perfectly sized.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::transformMatrixForProperty const):
    - Update for swapped apply arguments.

* Source/WebCore/rendering/mathml/RenderMathMLOperator.cpp:
    - Remove unused includes.

* Source/WebCore/rendering/style/RenderStyle.cpp:
    - Use existing accessors rather than writing full path out each time.
    - Use auto* rather than TransformOperation* to allow for de-virtualization
      of the calls to apply().

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
    - Take TransformOperations by r-value to avoid all sets causing a copy. Like
      with Length, StyleBuilder will do an explicit copy when forwarding.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::forwardInheritedValue):
    - Add overload of forwardInheritedValue for TransformOperations that copies
      before moving into RenderStyle.

* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasTransform const):
    - Call isEmpty() directly on the container.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Update to directly access operations via the member variable, necessary
      now that there is no operations() function.

* Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
(Nicosia::applyTransformAnimation):
    - Update for swapped apply arguments.

Canonical link: https://commits.webkit.org/279618@main
